spdx-utils/src/main/kotlin/SpdxExpression.kt

```diff
@@ -242,6 +242,17 @@ data class SpdxCompoundExpression(
             SpdxOperator.AND -> left.offersChoice() || right.offersChoice()
         }
 
+    override fun equals(other: Any?): Boolean {
```

Comment:
Why are you using the `equals` function for this? Did you consider `isEquivalentTo()?
...The contract is much simpler and debugging easier (for dedicated functions).

I think you should use a dedicated function.

Reply:
The reason was that my impression was that in all cases where we compare SPDX expressions we care about the semantics, but not about the structure. E.g. in a set of expressions, like used in `validChoices`, it's useful if you can rely on the fact that it does not contain multiple semantically equal expressions. Also when comparing two expressions using `==` I would expect that it is true if the semantics are the same, but not care about structure. Like if you write `1 + 1 == 2`.

Reply:
The reasons I asked for changing it is:
1. It can be extremely hard to figure out who is calling it, and thus it's similar hard to estimate the side effects of a change in its implementation
2. Equals return `true` for expressions which aren't equal: like `A OR A.equals(A) -> true` or `A OR B.equals(B OR A) -> true`. I can foresee that there is code caring about those differences, for example the code which determines whether two different expression trees are equivalent. Maybe we'll have code which cares about the order (for readability).
3. One is inheriting a complex contract: equals, hashcode, compareTo

After reading your arguments I'm not 100% convinced anymore that my proposal is better, so I'm ok if you keep it this way as well.

Reply:
> Equals return `true` for expressions which aren't equal

Well, they *are* equal *semantically*. So the question is: Do we expect `equals()` to check for semantical or syntactical equivalence, and do we need to be able to distinguish between the two. I basically agree that there might be cases where we would care about syntactical equivalence, e.g. to sort `OR`-ed licenses alphabetically for a canonical form to present to the user. In that case we would want to know the difference between `A OR B` and `B OR A`.

Also, [Kotlin docs](https://kotlinlang.org/docs/reference/equality.html) say for `equals()` to check about "structural equality", which to me is rather "syntactical equality" than "semantical equality".

That said, I'm yet undecided how to reflect these needs best. I somewhat like the `isEquivalentTo()` proposal, but "equivalent" and "equal" sound very similar, and it might be easy to confuse them.

Reply:
They don't really define what they mean by "structural equality", I think it's up to us to define what expressions we want to be equal.

I'm fine with moving this to a separate function, given that otherwise we would have to add a separate function to check if the expression trees are identical, once we need this. I will do this later today.

Reply:
But when not implementing semantic equality via `equals()`, wouldn't then most of the upcoming simplifications in this PR not be possible? Maybe it's still better to put the syntactic equality check into separate functions once we need it.

Reply:
As discussed offline I'll keep it as is. For the record, we agreed that in almost all cases we care about semantic equality, and it is very handy when comparing SPDX expressions with `==` or putting them in sets. If we need to compare syntactic equality, we can still implement another helper function.